### PR TITLE
[ML] Write header to autodetect before it is visible to other calls

### LINKED ISSUE #40385: ml/set_upgrade_mode.yml#"Setting upgrade_mode to enabled" failure on CI
Failed on 7.0 intake: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.0+intake/500/console (new intake commit is an unrelated docs change)

Reproduce line (does not reproduce locally):
```
./gradlew :x-pack:plugin:integTestRunner \
  -Dtests.seed=971A963D9AC18AC0 \
  -Dtests.class=org.elasticsearch.xpack.test.rest.XPackRestIT \
  -Dtests.method="test {p0=ml/set_upgrade_mode/Setting upgrade_mode to enabled}" \
  -Dtests.security.manager=true \
  -Dtests.locale=lt \
  -Dtests.timezone=Europe/Lisbon \
  -Dcompiler.java=12 \
  -Druntime.java=8
```

Stack trace:
```
java.lang.RuntimeException: Had to resort to force-closing jobs, something went wrong?
	at __randomizedtesting.SeedInfo.seed([971A963D9AC18AC0:1F4EA9E7343DE738]:0)
	at org.elasticsearch.xpack.core.ml.integration.MlRestTestStateCleaner.deleteAllJobs(MlRestTestStateCleaner.java:85)
	at org.elasticsearch.xpack.core.ml.integration.MlRestTestStateCleaner.clearMlMetadata(MlRestTestStateCleaner.java:31)
	at org.elasticsearch.xpack.test.rest.XPackRestIT.clearMlState(XPackRestIT.java:270)
	at org.elasticsearch.xpack.test.rest.XPackRestIT.cleanup(XPackRestIT.java:254)
	at sun.reflect.GeneratedMethodAccessor13.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
	at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
	at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894)
	at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
	at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.elasticsearch.client.ResponseException: method [POST], host [http://[::1]:38309], URI [/_ml/anomaly_detectors/_all/_close], status line [HTTP/1.1 500 Internal Server Error]
{"error":{"root_cause":[{"type":"exception","reason":"Unable to read field length from input stream\nFailed to parse length encoded header from stream\nFailed to handle all input data\nMl anomaly detector job failed\n"}],"type":"exception","reason":"Exception closing autodetect process","caused_by":{"type":"exception","reason":"Unable to read field length from input stream\nFailed to parse length encoded header from stream\nFailed to handle all input data\nMl anomaly detector job failed\n"}},"status":500}
	at org.elasticsearch.client.RestClient.convertResponse(RestClient.java:260)
	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:238)
	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:212)
	at org.elasticsearch.xpack.core.ml.integration.MlRestTestStateCleaner.deleteAllJobs(MlRestTestStateCleaner.java:77)
	... 37 more
```

### PR DESCRIPTION
**Pull request by davidkyle**
Write the header before publishing the process in the map so any calls to close, write etc cannot be made before (or during) the header is written. 

Closes #40385

Release Note: Closes a race condition where opening a job would fail with the error 'Failed to parse length encoded header from stream' if it was closed at the same time.

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java [hunk 1]
```diff
@@ -639,7 +643,7 @@ public void closeJob(JobTask jobTask, boolean restart, String reason) {
         processContext.tryLock();
         try {
             if (processContext.setDying() == false) {
-                logger.debug("Cannot close job [{}] as it has already been closed", jobId);
+                logger.debug("Cannot close job [{}] as it has been marked as dying", jobId);
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java]: This error message was the same as the one above so I change it.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java [hunk 2]
```diff
@@ -458,8 +458,11 @@ protected void doRun() {
 
                                     try {
                                         createProcessAndSetRunning(processContext, job, params, closeHandler);
-                                        processContext.getAutodetectCommunicator().init(params.modelSnapshot());
+                                        processContext.getAutodetectCommunicator().restoreState(params.modelSnapshot());
                                         setJobState(jobTask, JobState.OPENED);
+
+                                        assert processByAllocation.get(jobTask.getAllocationId()) == null;
+                                        processByAllocation.put(jobTask.getAllocationId(), processContext);
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java]: The fix for the bug is the `writeHeader` change but I thought moving this here would be a sensible change. 

Once added to the `processByAllocation` map the process can be used so why add `processContext` before it is fully formed? There are work arounds in the close() function for the case where `processContext` does not have an AutodetectCommunicator and I can't see the reason why it was not set here before.

Comment (droberts195) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java]: Before you made the job in index changes `processByAllocation.put(jobTask.getAllocationId(), processContext);` was called at the very beginning of `openJob` before any async actions were called.  Currently it's in between two async calls which I agree solves nothing.  But there may have originally been a good reason why it was done at the beginning of `openJob` before any async calls were made, for example to fix a race condition when two identical calls to `open_job` are made in very quick succession.  @dimitris-athanasiou can you remember?

**Review by droberts195 — COMMENTED** [review]
The change to call `writeHeader` in a different place is fine, but I think some of the other refactoring could be introducing race conditions and may also be highlighting that the job in index work has already introduced an unlikely race condition.

**Comment by davidkyle** [comment]
I reverted the refactoring as it caused tests to fail. There are improvements to be made here not least because the current position of `processByAllocation.putIfAbsent(jobTask.getAllocationId(), new ProcessContext(jobTask));`  appears to be due to an unrelated change rather than by design.

**Review by droberts195 — APPROVED** [review]
LGTM

**Comment by droberts195** [comment]
Another situation where data could be written to `autodetect` after the header prior to this change was as follows:

1. Open job in a single node cluster
2. Stop the node
3. Start the node - the job is still in the `opened` state, but doesn't immediately have a working `autodetect` process
4. Post data to the job - this can be done immediately after starting the node because the job is `opened`

**Pull request merged**